modules/infra/control/loopback.c

```diff
@@ -0,0 +1,196 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2024 Christophe Fontaine
+
+#include <gr.h>
+#include <gr_control_input.h>
+#include <gr_control_output.h>
+#include <gr_eth.h>
+#include <gr_iface.h>
+#include <gr_infra.h>
+#include <gr_log.h>
+#include <gr_loopback.h>
+#include <gr_mempool.h>
+#include <gr_module.h>
+
+#include <event2/event.h>
+
+#include <fcntl.h>
+#include <linux/if_tun.h>
+#include <net/if.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#define TUN_TAP_DEV_PATH "/dev/net/tun"
+
+static struct rte_mempool *loopback_pool;
+extern control_input_t control_to_loop_input;
+static struct event_base *global_ev_base;
+
+struct iface_info_loop {
+	int fd;
+	struct event *ev;
+};
+
+void loopback_tx(struct rte_mbuf *m) {
+	struct mbuf_data *d = mbuf_data(m);
+	struct iface_info_loop *loop;
+
+	loop = (struct iface_info_loop *)d->iface->info;
+	if (write(loop->fd, rte_pktmbuf_mtod(m, char *), m->data_len) < 0)
+		LOG(ERR, "write to tun device %s failed %s", d->iface->name, strerror(errno));
+
+	rte_pktmbuf_free(m);
+}
+
+static void iface_loopback_poll(evutil_socket_t, short, void *ev_iface) {
+	struct eth_input_mbuf_data *e;
+	struct iface_info_loop *loop;
+	struct iface *iface = ev_iface;
+	struct rte_mbuf *mbuf;
+	size_t len;
+	char *data;
+
+	loop = (struct iface_info_loop *)iface->info;
+
+	mbuf = rte_pktmbuf_alloc(loopback_pool);
+	if (!mbuf)
+		goto end;
+
+	mbuf->data_off = 256;
+	data = rte_pktmbuf_mtod(mbuf, char *);
+	len = read(loop->fd, data, 1600);
```

Comment:
Could we have a different MTU configured on the loopback from linux? Is there any way to detect this?

Reply:
yes, will fix it.